pdns/syncres.hh

```diff
@@ -214,60 +214,74 @@ private:
 template<class Thing> class Counters : public boost::noncopyable
 {
 public:
-  Counters()
-  {
+  typedef unsigned long counter_t;
+  struct value_t {
+    counter_t value;
+    time_t last;
+  };
+  typedef std::map<Thing, value_t> cont_t;
+
+  cont_t getMap() const {
+    return d_cont;
   }
-  unsigned long value(const Thing& t) const
+  counter_t value(const Thing& t) const
   {
-    typename cont_t::const_iterator i=d_cont.find(t);
+    typename cont_t::const_iterator i = d_cont.find(t);
 
-    if(i==d_cont.end()) {
+    if (i == d_cont.end()) {
       return 0;
     }
-    return (unsigned long)i->second;
+    return i->second.value;
   }
-  unsigned long incr(const Thing& t)
+
+  counter_t incr(const Thing& t, const struct timeval & now)
   {
-    typename cont_t::iterator i=d_cont.find(t);
+    typename cont_t::iterator i = d_cont.find(t);
 
-    if(i==d_cont.end()) {
-      d_cont[t]=1;
+    if (i == d_cont.end()) {
+      d_cont[t].value = 1;
+      d_cont[t].last = now.tv_sec;
```

Comment:
Might be worth taking a reference via `std::map::insert()` to avoid doing the lookup twice?